pkg/reconciler/source/config_watcher.go

```diff
@@ -20,143 +20,232 @@ import (
+	if dcmw, ok := cmw.(configmap.DefaultingWatcher); ok {
+		dcmw.WatchWithDefault(corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Name: cmName},
+			Data:       map[string]string{},
+		}, obs)
 
-		r.tracingCfg = tracingCfg
-		r.logger.Debug("Update from tracing ConfigMap", zap.Any("ConfigMap", cfg))
+	} else {
+		cmw.Watch(cmName, obs)
 	}
 }
 
-func (r *ConfigWatcher) LoggingConfig() *pkgLogging.Config {
-	if r == nil {
+// LoggingConfig returns the logging configuration from the ConfigWatcher.
+func (cw *ConfigWatcher) LoggingConfig() *pkglogging.Config {
+	if cw == nil {
 		return nil
 	}
-	return r.loggingConfig
+	return cw.loggingCfg
 }
 
-func (r *ConfigWatcher) MetricsConfig() *metrics.ExporterOptions {
-	if r == nil {
+// MetricsConfig returns the metrics configuration from the ConfigWatcher.
+func (cw *ConfigWatcher) MetricsConfig() *metrics.ExporterOptions {
+	if cw == nil {
 		return nil
 	}
-	return r.metricsConfig
+	return cw.metricsCfg
 }
 
-func (r *ConfigWatcher) TracingConfig() *tracingconfig.Config {
-	if r == nil {
+// TracingConfig returns the tracing configuration from the ConfigWatcher.
+func (cw *ConfigWatcher) TracingConfig() *tracingconfig.Config {
+	if cw == nil {
 		return nil
 	}
-	return r.tracingCfg
+	return cw.tracingCfg
+}
+
+func (cw *ConfigWatcher) updateFromLoggingConfigMap(cfg *corev1.ConfigMap) {
+	if cfg == nil {
+		return
+	}
+
+	delete(cfg.Data, "_example")
+
+	loggingCfg, err := pkglogging.NewConfigFromConfigMap(cfg)
+	if err != nil {
+		cw.logger.Warn("failed to create logging config from ConfigMap", zap.String("cfg.Name", cfg.Name))
+		return
+	}
+
+	cw.loggingCfg = loggingCfg
+
+	cw.logger.Debug("Updated logging config from ConfigMap", zap.Any("ConfigMap", cfg))
 }
 
+func (cw *ConfigWatcher) updateFromMetricsConfigMap(cfg *corev1.ConfigMap) {
+	if cfg == nil {
+		return
+	}
+
+	delete(cfg.Data, "_example")
+
+	cw.metricsCfg = &metrics.ExporterOptions{
+		Domain:    metrics.Domain(),
+		Component: cw.component,
+		ConfigMap: cfg.Data,
+	}
+
+	cw.logger.Debug("Updated metrics config from ConfigMap", zap.Any("ConfigMap", cfg))
+}
+
+func (cw *ConfigWatcher) updateFromTracingConfigMap(cfg *corev1.ConfigMap) {
+	if cfg == nil {
+		return
+	}
+
+	delete(cfg.Data, "_example")
+
+	tracingCfg, err := tracingconfig.NewTracingConfigFromMap(cfg.Data)
+	if err != nil {
+		cw.logger.Warn("failed to create tracing config from ConfigMap", zap.String("cfg.Name", cfg.Name))
+		return
+	}
+
+	cw.tracingCfg = tracingCfg
+
+	cw.logger.Debug("Updated tracing config from ConfigMap", zap.Any("ConfigMap", cfg))
+}
+
+// ToEnvVars serializes the contents of the ConfigWatcher to individual
+// environment variables.
 func (r *ConfigWatcher) ToEnvVars() []corev1.EnvVar {
-	loggingConfig, err := pkgLogging.LoggingConfigToJson(r.LoggingConfig())
+	envs := make([]corev1.EnvVar, 0, 3)
+
+	loggingCfg, err := pkglogging.LoggingConfigToJson(r.LoggingConfig())
```

Comment:
I can look into simplifying it.

Regarding the booleans, I understand the usage of pointers here as a way to distinguish between empty and unset value (that's one of the guidelines of the Kubernetes API btw.). How would you feel about initializing each config to an empty struct in the constructor, so that we always return at least an empty var for the required configs?

Reply:
That sounds reasonable too, as soon as it doesn't break the code :smile:

Reply:
Done in e640e85